### PR TITLE
XMR Scraper

### PR DESCRIPTION
**Pull request by joshpurvis**
This scraper is bootstrapped with hard coded values as of height 1654077. As it runs it keeps track of latest height's supply so that it only has to calculate the difference on each new block.

I left the docker-compose lines commented out for now, as I haven't built the docker image yet. I was already running the binary on my computer locally for testing. I did put in values that I thought should work based on my previous knowledge of docker and monero. I can look into this tomorrow if you need, and if you tell me where to push the image.

#23

### CODE REVIEW

File: blockchain-scrapers/docker-compose.yml [hunk 1]
```diff
@@ -193,6 +193,26 @@ services:
         max-size: "50m"
     secrets:
       - api_diadata
+# TODO: build docker image from https://github.com/monero-project/monero/blob/master/Dockerfile
+#  monero:
+#    image:
+#      diadata/monero-docker
```
Comment (frranck) [inline blockchain-scrapers/docker-compose.yml]: Could you uncomment and point to the right monero image ?

**Comment by frranck** [comment]
Hi Josh, you don't need to push any image, but I expect it to work with docker-compose.

**Comment by joshpurvis** [comment]
@frranck Is that what you mean? Sorry I just noticed that there were two types of images in the docker-compose.yml (blockchain nodes and the scraper docker images), I changed it to the latter.

**Comment by joshpurvis** [comment]
I'll see if I can get a monero fullnode dockerfile working for you a bit later today.

**Comment by joshpurvis** [comment]
@frranck -- Ok I've pushed a docker image for monero fullnode to my docker hub (https://hub.docker.com/r/joshpurvis/monero/). The source is available here if you'd like to maintain your own copy: https://github.com/joshpurvis/monero-docker

Also updated this PR with example docker-compose stanza for running it. 

Let me know if you have issues!

**Comment by frranck** [comment]
Hi, Could you copy your externa docker file  inside your PR ? we prefer to have it inside the repo. simply rename it and point to it inside the dockercompose.

**Comment by joshpurvis** [comment]
Sure, copied it.. I didn't do that because the other scrapers didn't have it (Cardano, for instance)

File: blockchain-scrapers/docker-compose.yml [hunk 2]
```diff
@@ -96,6 +96,25 @@ services:
       restart_policy:
         delay: 2s
         window: 20s
+  monero:
+    build:
```
Comment (frranck) [inline blockchain-scrapers/docker-compose.yml]: Now we're missing an image name:

    image:
      ${DOCKER_HUB_LOGIN}/monero

**Comment by frranck** [comment]
the image name should be ${DOCKER_HUB_LOGIN}/blockchain-scrapers_monero

**Comment by frranck** [comment]
so i could run it, but nothing is happening.
some

> Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 2018/09/06 08:34:03 Failed to retrieve current height from node
> 
> 
until I could get the monero node running, then it stalls.

if I restart it:

> 2018/09/06 09:05:19 Loaded secret in /run/secrets/api_diadata

**Comment by joshpurvis** [comment]
It stalls because the current height is less than the snapshot height. It only prints when it gets a new height. 

I can either add a message saying it's "waiting for the blockchain to sync", or could change it so it publishes a number every 10 seconds regardless if it's new or not. Which would you prefer?

Also, raises another question, do you want it to publish a full replay of the historic supply, or just new values after the snapshot?

**Comment by frranck** [comment]
There's something in the api https://github.com/diadata-org/api-golang/blob/master/dia/ApiClient.go#L162 to deal with not flooding the api, but still sending at least every hour. but we need to send it always when the service starts.

**Comment by joshpurvis** [comment]
Ok I've changed it so that it always publishes latest available height -- in the case of the blockchain being out of sync, it's the snapshot value:

```
2018/09/06 11:19:45 Using snapshot values which are as of height 1654077
2018/09/06 11:19:45 Updating circulating supply as of block 1654077: 16377801.401737
2018/09/06 11:19:45 Waiting for blockchain to sync...
```

It'll do that every 10 seconds until synced is greater than snapshot values.

After sync is greater than snapshot, it starts calculating and publishing the supply of new heights as they're achieved by monerod. Also sending those every 10 seconds.

**Comment by frranck** [comment]
I'm not sure I understand, I don't want any outdated value when it's synching.

and the image needs to be called like that:
  monero:
    image:
      ${DOCKER_HUB_LOGIN}/blockchain-scrapers_monero

**Comment by joshpurvis** [comment]
> I'm not sure I understand, I don't want any outdated value when it's synching.

You said you wanted it to publish a value as soon as it starts:

> but we need to send it always when the service starts.

If the blockchain its querying is out of sync, it can only send an outdated value. 

If tell me the behavior you want to see, we can probably work out another way, but your requirements seem ambiguous to me. (unless I'm misunderstanding something?)

> and the image needs to be called like that:
> monero:
> image:
> ${DOCKER_HUB_LOGIN}/blockchain-scrapers_monero

^ Ok changed it to that, instead of what was previously given

**Comment by frranck** [comment]
> You said you wanted it to publish a value as soon as it starts:
Yes but I want a correct value, not any value...
if you have a value, you send it even if it was sent before, 
if you don't you wait until you have one.
 the refresh/don't send twice behavior is already coded there https://github.com/diadata-org/api-golang/blob/master/dia/ApiClient.go#L163
so sum up: send values as much as you want, but correct ones ;)

**Comment by joshpurvis** [comment]
If I understand you correctly, that's how it works now on this latest iteration.. It would never send a wrong value, but it could send an old value if the monero blockchain isn't synchronized for whatever reason. (since supply is calculated dynamically from available blockchain data)

Basically if you turn it on from scratch like you did the other day (no blockchain syncing has happened yet), it will:
1. Start publishing the latest thing it knows (the supply as of the snapshot I hardcoded at height 1654077). It will do this every 10 seconds until the blockchain sync'd height is greater than that snapshot height.
2. Once sync'd height > snapshot height, it will start calculating and publishing those new heights until it reaches the present height.
3. Finally it will continue to calculate and publish the new total supply every 10 seconds. If no new block has been found it might republish the supply from the last known height. (I assume that throttling code you linked to should prevent this from spamming too much)
4. If the scraper is restarted for some reason it will query the blockchain for the latest height and publish the supply as of that height and then continue updating from there forward.

**Comment by frranck** [comment]
@joshpurvis  Can you know how old is the last block you synched ?

**Comment by joshpurvis** [comment]
Yes, I believe so.. by getting the block header, you can get the timestamp of that block

**Comment by frranck** [comment]
Then  can't we figure out we're synching by filtering out blocks that out too old ? we're doing this for BTC and LTC using a constant: https://github.com/diadata-org/api-golang/blob/master/blockchain-scrapers/cmd/btc/btc.go#L14

**Comment by frranck** [comment]
@joshpurvis Yes It's perfect, it's just the first run since we have persistency.

**Comment by frranck** [comment]
Ok thanks for your PR , I merged it there: https://github.com/diadata-org/api-golang/pull/41

**Pull request closed without merging**